### ISSUE TITLE
cmd: relocate script test framework from cmd/go to cmd/internal

### ISSUE THREAD

**Issue by thanm** [golang/go#68606]
The Go command has a very nice "script" testing framework that makes it easy to write regression tests that involve smallish multifile test cases and sequences of "go" command invocations. Other tools (compiler, linker, etc) would benefit from a similar facility. The convention for the compiler has been to add new test cases in $GOROOT/test, which then run by the harness in cmd/internal/testdir, howewer the harness code there has a lot of peculiarities: except for the simplest cases (e.g. compiler crash when run on a single file) it is difficult for test writers to pick out the correct flavor of test (rundir, runindir, compiledir, etc) to get the behavior they want.

To make it easier and simpler to code up more complicated test cases, I propose to relocate the cmd/go script test framework from cmd/go to cmd/internal, then make it available in some form for testing our various tools (compile, link, nm, etc). This is a tracking issue to record the progress on this effort.

**Comment by cuonglm** [comment]
When seeing "script", I thought it's the `go test -run=TestScript`, which involves writing test case using https://pkg.go.dev/golang.org/x/tools@v0.23.0/txtar

Either way, it would be a great improvement.

**Comment by thanm** [comment]
> When seeing "script", I thought it's the `go test -run=TestScript`, which involves writing test case using https://pkg.go.dev/golang.org/x/tools@v0.23.0/txtar

Yes, the intent is to support txtar-style tests, such as this one in cmd/go: 

https://go.googlesource.com/go/+/b788e91badd523e5bb0fc8d50cd76b8ae04ffb20/src/cmd/go/testdata/script/link_external_undef.txt

**Comment by thanm** [comment]
Closing out this issue, the new framework is running + functional. Happy script test writing...
